src/ucp/wireup/wireup.c

```diff
@@ -975,6 +967,65 @@ ucp_wireup_get_reachable_mds(ucp_worker_h worker,
     key->reachable_md_map = dst_md_map;
 }
 
+static void
+ucp_wireup_reclaim_unused_uct_ep(ucp_ep_h ep, ucp_ep_config_key_t *new_key,
+                                 ucs_queue_head_t *uct_ep_req_queue)
+{
+    uct_ep_h new_uct_eps[UCP_MAX_LANES]            = {NULL};
+    ucp_lane_index_t reuse_lane_map[UCP_MAX_LANES] = {UCP_NULL_LANE};
+    ucp_ep_config_key_t *old_key                   = &ucp_ep_config(ep)->key;
+    ucp_lane_index_t lane, reuse_lane;
+
+    ucs_assert(new_key->num_lanes > 1);
+    if ((ep->cfg_index == 0) || (old_key->num_lanes == new_key->num_lanes)) {
+        return;
+    }
+
+    for (lane = 0; lane < old_key->num_lanes; ++lane) {
```

Comment:
not sure we need all this logic here..
I would expect we can just go over all lanes between old_key->num_lanes and new_key->num_lanes-1 and discard them. this would be either 0 lanes [when symmetric config] or 1 lane [when asymmetric config].
also can ucs_assertv(new_key->num_lanes >= old_key->num_lanes, "...", ...)

Reply:
@yosefe Do you mean below logic for integration3 branch since we don't need to consider multi devices?
```
static void 
ucp_wireup_reclaim_unused_uct_ep(ucp_ep_h ep, ucp_ep_config_key_t *new_key,
                                 ucs_queue_head_t *uct_ep_req_queue)
{
    ucp_ep_config_key_t *old_key  = &ucp_ep_config(ep)->key;
    ucp_lane_index_t lane;

    ucs_assert(new_key->num_lanes > 1);
    if ((ep->cfg_index == 0) || (old_key->num_lanes == new_key->num_lanes)) {
        return;
    }    

    ucs_assertv(old_key->num_lanes > new_key->num_lanes,
                "inited lane : %d, selected lane : %d", 
                (uint32_t)(old_key->num_lanes), (uint32_t)(new_key->num_lanes));
    for (lane = new_key->num_lanes; lane < old_key->num_lanes; ++lane) {
        uct_ep_pending_purge(ep->uct_eps[lane], ucp_wireup_pending_purge_cb,
                             uct_ep_req_queue);
        uct_ep_destroy(ep->uct_eps[lane]);
        ep->uct_eps[lane] = NULL;
    }    

    return;
}
```

Reply:
yes, something like that. could also remove (old_key->num_lanes == new_key->num_lanes) from the if condition.
does it work?

Reply:
It works when 1) ```server: NUM_PATH=1; client: NUM_PATH=2``` 2) ```server: NUM_PATH=2; Client: NUM_PATH=2``` 3) ```server: NUM_PATH=2; Client: NUM_PATH=1``` 4) ```server: NUM_PATH=1; Client: NUM_PATH=1```

Reply:
@yosefe Done